model/docs/extending_pyrenew.qmd

```diff
@@ -0,0 +1,256 @@
+---
+title: Extending pyrenew
+format: gfm
+---
+
+This tutorial illustrates how to extend `pyrenew` with custom `RandomVariable` classes. We will use the `InfectionsWithFeedback` class as an example. The `InfectionsWithFeedback` class is a `RandomVariable` that models the number of infections at time $t$ as a function of the number of infections at time $t - \tau$ and the reproduction number at time $t$. The reproduction number at time $t$ is a function of the *unadjusted* reproduction number at time $t - \tau$ and the number of infections at time $t - \tau$:
+
+$$
+\begin{align*}
+I(t) & = \mathcal{R}(t)\sum_{\tau=1}^{T_g}I(t - \tau)g(\tau) \\
+\mathcal{R}(t) & = \mathcal{R}^u(t)\exp\left(-\gamma(t)\sum_{\tau=1}^{T_f}I(t - \tau)f(\tau)\right)
+\end{align*}
+$$
+
+Where $\mathcal{R}^u(t)$ is the unadjusted reproduction number, $g(t)$ is the generation interval, $\gamma(t)$ is the infection feedback strength, and $f(t)$ is the infection feedback pmf.
+
+## The expected outcome
+
+Before we start, let's simulate the model with the original `InfectionsWithFeedback` class. We will simulate the model with no observation process. The following code-chunk loads the required libraries and defines the model components:
+
+```{python}
+#| label: setup
+import jax
+import jax.numpy as jnp
+import numpy as np
+import numpyro as npro
+import numpyro.distributions as dist
+from pyrenew.deterministic import DeterministicPMF, DeterministicVariable
+from pyrenew.latent import InfectionsWithFeedback
+from pyrenew.model import RtInfectionsRenewalModel
+from pyrenew.process import RtRandomWalkProcess
+from pyrenew.metaclass import DistributionalRV
+```
+
+The following code-chunk defines the model components. Notice that for both the generation interval and the infection feedback, we use a deterministic PMF with equal probabilities:
+
+```{python}
+#| label: model-components
+gen_int = DeterministicPMF(jnp.array([0.25, 0.5, 0.15, 0.1]))
+
+I0 = DistributionalRV(dist=dist.LogNormal(0, 1), name="I0")
+
+latent_infections = InfectionsWithFeedback(
+  infection_feedback_strength = DeterministicVariable(1.1),
+  infection_feedback_pmf = gen_int,
+)
+
+rt = RtRandomWalkProcess()
+```
+
+With all the components defined, we can build the model:
+
+```{python}
+#| label: build1
+model0 = RtInfectionsRenewalModel(
+    gen_int=gen_int,
+    I0=I0,
+    latent_infections=latent_infections,
+    Rt_process=rt,
+    observation_process=None,
+)
+```
+
+And simulate it:
+
+```{python}
+#| label: simulate1
+# Sampling and fitting model 0 (with no obs for infections)
+np.random.seed(223)
+with npro.handlers.seed(rng_seed=np.random.randint(1, 600)):
+    model0_samp = model0.sample(n_timepoints=30)
```

Comment:
If we specify 30 timepoints, why does the plot only show 18?

Reply:
B/c the rest go to inf! Changed the feedback to avoid this.

Reply:
Addressed in 5eb31cc